feat: deprecate TRANSFORMERS_CACHE, use HF_HUB_CACHE everywhere

tjohnson31415 opened the pull request:
#### Motivation

`TRANSFORMERS_CACHE` is deprecated (slated for removal with Transformers v5) and `HUGGINGFACE_HUB_CACHE` is legacy. This PR standardizes on `HF_HUB_CACHE` to configure the cache. Also, not all operations/CLI commands were correctly pulling from `TRANSFORMERS_CACHE` so we have been setting both env vars anyways. After this change, everything should work with only `HF_HUB_CACHE`.

#### Modifications

- Launcher inspects HF_HUB_CACHE to determine the model cache path
    - TRANSFORMERS_CACHE and HUGGINGFACE_HUB_CACHE are still checked as well, but a deprecation warning is printed
    - if multiple values are present and do not match, raise an error
- Launcher can resolve the default HF_HUB_CACHE so it does not need to be set (HF_HOME or its default can be used instead)
- Server CLI checks TRANSFORMERS_CACHE and prints a warning if it is set
- Server CLI returns an error if both TRANSFORMERS_CACHE and HF_HUB_CACHE are set with different values

#### Result

[Describe how the changes affects existing behavior and how to test it]

njhill commented on launcher/src/main.rs:
```diff
@@ -139,7 +139,55 @@ fn main() -> ExitCode {
+    let mut cache_env_var: String = "".to_string();
+    let mut cache_path: String = "".to_string();
+
+    match env::var("HF_HUB_CACHE") {
+        Ok(t) => {
+            cache_env_var = "HF_HUB_CACHE".into();
+            cache_path = t.into();
+        },
+        _ => {}
+    }
```
nit: Simpler as `if let Ok(t) = env::var("HF_HUB_CACHE") {` ...?

njhill commented on launcher/src/main.rs:
```diff
@@ -139,7 +139,55 @@ fn main() -> ExitCode {
+    for deprecated_env_var in vec!["TRANSFORMERS_CACHE", "HUGGINGFACE_HUB_CACHE"] {
+        match (
+            env::var(deprecated_env_var),
+            cache_env_var.len() > 0,
```
nit:
```suggestion
            !cache_env_var.is_empty(),
```
?

njhill commented on launcher/src/main.rs:
```diff
@@ -139,7 +139,55 @@ fn main() -> ExitCode {
+    // ensure HF_HUB_CACHE is set for downstream usage
+    // default value to match huggingface_hub
+    // REF: https://github.com/huggingface/huggingface_hub/blob/5ff2d150d121d04799b78bc08f2343c21b8f07a9/docs/source/en/package_reference/environment_variables.md?plain=1#L32
+    if cache_path.len() == 0 {
```
nit:
```suggestion
    if cache_path.is_empty() {
```
?

njhill commented on launcher/src/main.rs:
```diff
@@ -139,7 +139,55 @@ fn main() -> ExitCode {
+    // default value to match huggingface_hub
+    // REF: https://github.com/huggingface/huggingface_hub/blob/5ff2d150d121d04799b78bc08f2343c21b8f07a9/docs/source/en/package_reference/environment_variables.md?plain=1#L32
+    if cache_path.len() == 0 {
+        if let Some(hf_home) = env::var("HF_HOME").ok() {
+            cache_path = hf_home + "/hub".into();
+        } else {
+            cache_path = "~/.cache/huggingface/hub".into();
+        }
```
nit suggestion:
```suggestion
        cache_path = if let Ok(hf_home) = env::var("HF_HOME") {
             hf_home + "/hub".into()
        } else {
             "~/.cache/huggingface/hub".into()
        }
```

njhill reviewed (approved):
This looks great thanks @tjohnson31415. Just had some minor rust nits.

tjohnson31415 commented:
@njhill I tested from a built image and found that using the `~` for the home directory does not resolve on its own. I adjusted the cache path defaulting to allow pulling from `$HOME` instead and refactored it to use proper path construction with `PathBuf` (instead of just string joins). PTAL.

njhill reviewed (approved):
Thanks @tjohnson31415 this was more work to do properly than I'd expected (what's new lol)!
